File: android/app/src/main/java/com/worktime/android/data/model/ReadModels.kt
package com.worktime.android.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class DashboardResponse(
    @SerialName("as_of") val asOf: String,
    val identity: Identity,
    @SerialName("work_context") val workContext: WorkContext,
    @SerialName("current_status") val currentStatus: CurrentStatus,
    @SerialName("next_shifts") val nextShifts: NextShifts,
    @SerialName("team_status") val teamStatus: TeamStatus,
    @SerialName("time_off_summary") val timeOffSummary: TimeOffSummary,
)

@Serializable
data class Identity(
    val id: Int,
    val username: String,
    @SerialName("display_name") val displayName: String,
    @SerialName("is_admin") val isAdmin: Boolean,
)

@Serializable
data class FeatureFlags(
    @SerialName("time_off_enabled") val timeOffEnabled: Boolean,
)

@Serializable
data class WorkContext(
    @SerialName("schedule_type") val scheduleType: String? = null,
    @SerialName("team_number") val teamNumber: Int? = null,
    @SerialName("effective_team_number") val effectiveTeamNumber: Int? = null,
    val state: String,
    @SerialName("feature_flags") val featureFlags: FeatureFlags,
)

@Serializable
data class ShiftSummary(
    val code: String,
    @SerialName("display_code") val displayCode: String,
    val name: String,
    @SerialName("start_hour") val startHour: Double? = null,
    @SerialName("end_hour") val endHour: Double? = null,
    @SerialName("is_working") val isWorking: Boolean,
)

@Serializable
data class TeamShiftStatus(
    @SerialName("team_number") val teamNumber: Int,
    val date: String,
    @SerialName("shift_day") val shiftDay: String,
    @SerialName("shift_code") val shiftCode: String,
    val shift: ShiftSummary,
    @SerialName("is_currently_working") val isCurrentlyWorking: Boolean,
)

@Serializable
data class OffDayProgress(
    val current: Int,
    val total: Int,
)

@Serializable
data class CurrentStatus(
    @SerialName("as_of") val asOf: String,
    @SerialName("current_shift") val currentShift: TeamShiftStatus? = null,
    @SerialName("currently_working_team") val currentlyWorkingTeam: TeamShiftStatus? = null,
    @SerialName("off_day_progress") val offDayProgress: OffDayProgress? = null,
)

@Serializable
data class NextShiftItem(
    @SerialName("team_number") val teamNumber: Int,
    val date: String,
    @SerialName("shift_code") val shiftCode: String,
    val shift: ShiftSummary,
)

@Serializable
data class NextShifts(
    @SerialName("as_of") val asOf: String,
    val items: List<NextShiftItem>,
)

@Serializable
data class TeamStatus(
    @SerialName("as_of") val asOf: String,
    val items: List<TeamShiftStatus>,
)

@Serializable
data class TimeOffSummaryItem(
    @SerialName("entry_id") val entryId: String,
    @SerialName("entry_kind") val entryKind: String,
    @SerialName("entry_type") val entryType: String,
    @SerialName("entry_flag") val entryFlag: String,
    val note: String? = null,
    @SerialName("starts_on") val startsOn: String,
    @SerialName("ends_on") val endsOn: String,
    @SerialName("is_recurring_weekly") val isRecurringWeekly: Boolean,
    @SerialName("is_active") val isActive: Boolean,
)

@Serializable
data class TimeOffSummary(
    @SerialName("as_of") val asOf: String,
    @SerialName("active_items") val activeItems: List<TimeOffSummaryItem>,
    @SerialName("upcoming_items") val upcomingItems: List<TimeOffSummaryItem>,
    @SerialName("total_upcoming") val totalUpcoming: Int,
)

@Serializable
data class TaskMutationRequest(
    val text: String? = null,
    @SerialName("label_id") val labelId: String? = null,
    @SerialName("start_time") val startTime: String? = null,
    @SerialName("stop_time") val stopTime: String? = null,
    @SerialName("includes_break") val includesBreak: Boolean? = null,
)

@Serializable
data class TaskRecord(
    val id: String,
    @SerialName("user_id") val userId: Int,
    @SerialName("label_id") val labelId: String? = null,
    val text: String,
    @SerialName("start_time") val startTime: String,
    @SerialName("stop_time") val stopTime: String? = null,
    @SerialName("includes_break") val includesBreak: Boolean,
    @SerialName("created_at") val createdAt: String,
)

@Serializable
data class WorkLocationMutationRequest(
    val date: String,
    @SerialName("country_code") val countryCode: String,
    val label: String? = null,
)

@Serializable
data class WorkLocationRecord(
    val id: Int,
    @SerialName("user_id") val userId: Int,
    val date: String,
    @SerialName("country_code") val countryCode: String,
    val label: String? = null,
    @SerialName("created_at") val createdAt: String,
)

@Serializable
data class WorkLocationListResponse(
    val items: List<WorkLocationRecord>,
    val total: Int,
)

@Serializable
data class SyncStatusResponse(
    @SerialName("labels_updated_at") val labelsUpdatedAt: String? = null,
    @SerialName("tasks_updated_at") val tasksUpdatedAt: String? = null,
    @SerialName("templates_updated_at") val templatesUpdatedAt: String? = null,
    @SerialName("work_locations_updated_at") val workLocationsUpdatedAt: String? = null,
    @SerialName("time_off_entries_updated_at") val timeOffEntriesUpdatedAt: String? = null,
    @SerialName("gantt_tasks_updated_at") val ganttTasksUpdatedAt: String? = null,
    @SerialName("preferences_updated_at") val preferencesUpdatedAt: String? = null,
    @SerialName("server_timestamp") val serverTimestamp: String,
)

File: backend/app/read_models.py
"""Reusable read models for mobile and other read-only clients."""

from __future__ import annotations

from datetime import date as dt_date
from datetime import datetime as dt_datetime
from typing import Literal

from pydantic import BaseModel

ScheduleType = Literal["9-5", "2-shift", "weekend-shift", "5-shift"]
ReadModelState = Literal["ready", "missing_work_context"]


class ReadModelIdentity(BaseModel):
    """Authenticated user identity exposed to read-only clients."""

    id: int
    username: str
    display_name: str
    is_admin: bool


class ReadModelFeatureFlags(BaseModel):
    """Feature flags relevant to read-only companion clients."""

    time_off_enabled: bool


class ReadModelWorkContext(BaseModel):
    """Persisted work context derived from synced user preferences."""

    schedule_type: ScheduleType | None
    team_number: int | None
    effective_team_number: int | None
    state: ReadModelState
    feature_flags: ReadModelFeatureFlags


class ReadModelShift(BaseModel):
    """Stable shift description used across multiple read models."""

    code: str
    display_code: str
    name: str
    start_hour: float | None
    end_hour: float | None
    is_working: bool


class OffDayProgressReadModel(BaseModel):
    """Progress through the current off-day block."""

    current: int
    total: int


class TeamShiftStatusReadModel(BaseModel):
    """Shift status for a single team."""

    team_number: int
    date: dt_date
    shift_day: dt_date
    shift_code: str
    shift: ReadModelShift
    is_currently_working: bool


class CurrentStatusReadModel(BaseModel):
    """Current shift status for the authenticated user's work context."""

    as_of: dt_datetime
    current_shift: TeamShiftStatusReadModel | None
    currently_working_team: TeamShiftStatusReadModel | None
    off_day_progress: OffDayProgressReadModel | None


class NextShiftItemReadModel(BaseModel):
    """Single upcoming shift entry."""

    team_number: int
    date: dt_date
    shift_code: str
    shift: ReadModelShift


class NextShiftsReadModel(BaseModel):
    """Ordered upcoming shifts for the authenticated user's team."""

    as_of: dt_datetime
    items: list[NextShiftItemReadModel]


class TeamStatusReadModel(BaseModel):
    """Current or nearby team status for all teams in the schedule."""

    as_of: dt_datetime
    items: list[TeamShiftStatusReadModel]


class TimeOffSummaryItemReadModel(BaseModel):
    """Upcoming or active time-off occurrence."""

    entry_id: str
    entry_kind: str
    entry_type: str
    entry_flag: str
    note: str | None
    starts_on: dt_date
    ends_on: dt_date
    is_recurring_weekly: bool
    is_active: bool


class TimeOffSummaryReadModel(BaseModel):
    """Upcoming and active time-off summary."""

    as_of: dt_datetime
    active_items: list[TimeOffSummaryItemReadModel]
    upcoming_items: list[TimeOffSummaryItemReadModel]
    total_upcoming: int


class DashboardReadModel(BaseModel):
    """Aggregated reusable read models for companion clients."""

    as_of: dt_datetime
    identity: ReadModelIdentity
    work_context: ReadModelWorkContext
    current_status: CurrentStatusReadModel
    next_shifts: NextShiftsReadModel
    team_status: TeamStatusReadModel
    time_off_summary: TimeOffSummaryReadModel
